### PR TITLE
[SPARK-33971][SQL] Eliminate distinct from more aggregates

### PR DESCRIPTION
**Pull request by tanelk**
### What changes were proposed in this pull request?

Add more aggregate expressions to `EliminateDistinct` rule.

### Why are the changes needed?

Distinct aggregation can add a significant overhead. It's better to remove distinct whenever possible.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UT

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -349,11 +349,20 @@ abstract class Optimizer(catalogManager: CatalogManager)
...
+  }
+
+  private def isDuplicateAgnostic(af: AggregateFunction): Boolean = af match {
+    case _: Max => true
+    case _: Min => true
+    case _: BitAndAgg => true
+    case _: BitOrAgg => true
+    case _: First => true
+    case _: Last => true
+    case _: HyperLogLogPlusPlus => true
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:363]: Can you show or prove that the duplication doesn't matter in `HyperLogLogPlusPlus`?

**Comment by HyukjinKwon** [comment]
Looks making sense otherwise. cc @gengliangwang and @cloud-fan FYI

**Comment by cloud-fan** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 1 above)
Comment (tanelk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:363]: There was a similar discussion before: https://github.com/apache/spark/pull/29740#discussion_r487692977

In short HLL++ keeps track of the max number of leading zeros in the hashes of inputs. Repeatedly adding the same valus does not change its results. But as @maropu said the last time, that perhaps we are not making any guarantees about it and so it might not be the best idea to include it here.

I'm okay with removing it if you say so.

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:363]: yea, removing it for now looks fine to me.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateDistinctSuite.scala [hunk 2]
```diff
@@ -32,25 +34,25 @@ class EliminateDistinctSuite extends PlanTest {
...
-      .select(minDistinct('a).as('result))
-      .analyze
-    val answer = testRelation
-      .select(min('a).as('result))
-      .analyze
-    assert(query != answer)
-    comparePlans(Optimize.execute(query), answer)
+  Seq(
+    ("max", Max(_)),
+    ("min", Min(_)),
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateDistinctSuite.scala:39]: We need the first name param? How about using `prettyName` instead?

**Review by maropu — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 3]
```diff
@@ -349,11 +349,19 @@ abstract class Optimizer(catalogManager: CatalogManager)
...
+    case ae: AggregateExpression if ae.isDistinct && isDuplicateAgnostic(ae.aggregateFunction) =>
+      ae.copy(isDistinct = false)
+  }
+
+  private def isDuplicateAgnostic(af: AggregateFunction): Boolean = af match {
+    case _: Max => true
+    case _: Min => true
+    case _: BitAndAgg => true
+    case _: BitOrAgg => true
+    case _: First => true
```
Comment (zhengruifeng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:361]: `First` and `Last` are non-deterministic, does this matter?

Comment (tanelk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:361]: In another PR I try to argue that they should be deterministic: #29810
TLDR: An analogous aggregator would be  sum on float and double datatype - its result does depend on the order of its inputs, but is deterministic.

Comment (zhengruifeng) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:361]: If we finally make `First` and `Last` deterministic, then I guess they need to be removed from `EliminateDistinct`?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:361]: Let's exclude First/Last here before #29810 is merged

**Comment by maropu** [comment]
@tanelk ping.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 3 above)
Comment (tanelk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Done

**Comment by maropu** [comment]
retest this please

**Review by maropu — APPROVED** [review]

**Review by zhengruifeng — APPROVED** [review]

**Comment by maropu** [comment]
Thanks! Merged to master.
